webknossos-tracingstore/app/com/scalableminds/webknossos/tracingstore/tracings/volume/MergedVolume.scala

```diff
@@ -0,0 +1,135 @@
+
+  def addLabelSetFromDataZip(zipFile: File)(implicit ec: ExecutionContext): Box[Unit] = {
+    val importLabelSet: mutable.Set[UnsignedInteger] = scala.collection.mutable.Set()
+    val unzipResult = withBucketsFromZip(zipFile) { (_, bytes) =>
+      val dataTyped =
+        UnsignedIntegerArray.fromByteArray(bytes, elementClass)
+      val nonZeroData = UnsignedIntegerArray.filterNonZero(dataTyped)
+      importLabelSet ++= nonZeroData
+    }
+    for {
+      _ <- unzipResult
+      _ = addLabelSet(importLabelSet)
+    } yield ()
+  }
+
+  def addLabelSetFromBucketStream(bucketStream: Iterator[(BucketPosition, Array[Byte])],
+                                  allowedResolutions: Set[Point3D]): Unit = {
+    val labelSet: mutable.Set[UnsignedInteger] = scala.collection.mutable.Set()
+    bucketStream.foreach {
+      case (bucketPosition, data) =>
+        if (allowedResolutions.contains(bucketPosition.resolution)) {
+          val dataTyped = UnsignedIntegerArray.fromByteArray(data, elementClass)
+          val nonZeroData: Array[UnsignedInteger] = UnsignedIntegerArray.filterNonZero(dataTyped)
+          labelSet ++= nonZeroData
+        }
+    }
+    addLabelSet(labelSet)
+  }
+
+  def addLabelSet(labelSet: mutable.Set[UnsignedInteger]): Unit = labelSets += labelSet
+
+  private def prepareLabelMaps(): Unit =
+    if (labelSets.isEmpty || (labelSets.length == 1 && initialLargestSegmentId == 0) || labelMaps.nonEmpty) {
+      ()
+    } else {
+      var i: UnsignedInteger = UnsignedInteger.zeroFromElementClass(elementClass)
+      if (initialLargestSegmentId > 0) {
+        labelMaps += mutable.HashMap.empty[UnsignedInteger, UnsignedInteger]
+        i = UnsignedInteger.fromLongWithElementClass(initialLargestSegmentId, elementClass)
+      }
+      labelSets.toList.foreach { labelSet =>
+        var labelMap = mutable.HashMap.empty[UnsignedInteger, UnsignedInteger]
+        labelSet.foreach { label =>
+          i = i.increment
+          labelMap += ((label, i))
+        }
+        labelMaps += labelMap
+      }
+      largestSegmentId = i
+    }
+
+  def addFromBucketStream(sourceVolumeIndex: Int,
+                          bucketStream: Iterator[(BucketPosition, Array[Byte])],
+                          allowedResolutions: Option[Set[Point3D]] = None): Unit =
+    bucketStream.foreach {
+      case (bucketPosition, bytes) =>
+        if (!isAllZero(bytes) && allowedResolutions.forall(_.contains(bucketPosition.resolution))) {
+          add(sourceVolumeIndex, bucketPosition, bytes)
+        }
+    }
+
+  def addFromDataZip(sourceVolumeIndex: Int, zipFile: File)(implicit ec: ExecutionContext): Box[Unit] =
+    withBucketsFromZip(zipFile) { (bucketPosition, bytes) =>
+      add(sourceVolumeIndex, bucketPosition, bytes)
+    }
+
+  def add(sourceVolumeIndex: Int, bucketPosition: BucketPosition, data: Array[Byte]): Unit =
+    if (data.length > 1) { // skip reverted buckets
```

Comment:
One question: do we need to keep this for backwards compatibility with old downloaded tracings because the iterators no longer return reverted buckets. Maybe we can put this check into the `addFromDataZip` call then.

Reply:
true enough :) and the data zip unpacking method drops all-zero buckets already also :) I’ll remove it here.